File: web-backend/internal/facts/envelope.go
// Package facts defines the audited market-data envelope shared by internal
// transports. It contains no MCP or HTTP behavior.
package facts

import "encoding/json"

// Envelope mirrors the Python ToolEnvelope produced by the audited data core.
// Market data and audit evidence stay as raw JSON so Go never rewrites them.
type Envelope struct {
	SchemaVersion string          `json:"schema_version"`
	RequestID     string          `json:"request_id"`
	Tool          string          `json:"tool"`
	OK            bool            `json:"ok"`
	Data          json.RawMessage `json:"data"`
	Sources       json.RawMessage `json:"sources"`
	DataAudit     json.RawMessage `json:"data_audit"`
	DataWarnings  json.RawMessage `json:"data_warnings"`
	DataPolicy    json.RawMessage `json:"data_policy"`
	QueriedAt     string          `json:"queried_at"`
	Error         *ToolError      `json:"error"`
}

// ToolError is the typed portion of an audited envelope error.
type ToolError struct {
	Code      string          `json:"code"`
	Message   string          `json:"message"`
	Retryable bool            `json:"retryable"`
	Details   json.RawMessage `json:"details"`
}

File: src/fund_advisor_data_core/contracts/envelope.py
"""ToolEnvelope contract shared across product boundaries."""

from __future__ import annotations

from datetime import datetime
from typing import Any, Literal
from uuid import UUID, uuid4

from pydantic import BaseModel, ConfigDict, Field

from .tools import ToolName


class ToolError(BaseModel):
    model_config = ConfigDict(extra="forbid")

    code: str
    message: str
    retryable: bool = False
    details: dict[str, Any] = Field(default_factory=dict)


class ToolEnvelope(BaseModel):
    model_config = ConfigDict(extra="forbid")

    schema_version: Literal["1.0"] = "1.0"
    request_id: UUID = Field(default_factory=uuid4)
    tool: ToolName
    ok: bool
    data: dict[str, Any] | None = None
    sources: list[dict[str, Any]] = Field(default_factory=list)
    data_audit: list[dict[str, Any]] = Field(default_factory=list)
    data_warnings: list[dict[str, Any] | str] = Field(default_factory=list)
    data_policy: dict[str, Any] = Field(default_factory=dict)
    queried_at: datetime
    error: ToolError | None = None

